### PR TITLE
Major update: new name, new fluent decorator api, isolated instances, label provider

### LINKED ISSUE #23: 100% code coverage for constraint decorators
All constraint decorators should have unit tests.

### LINKED ISSUE #24: Add Joi documentation as jsdoc comments to constraint decorators
All decorators should have jsdoc comments, describing how to use each decorator. This information should be duplicated from the Joi documentation, ideally with examples updated to use decorators*.

* Examples may not be necessary if we have unit tests demonstrating how to use each decorator: see #23.

### LINKED ISSUE #40: Label provider function
Make creating user friendly property labels easier (i.e. less code) for developers.

Joi's error messages are mostly user friendly except for the fact that they are based on property keys. A consumer can specify user friendly labels for properties via the `Label` decorator which is good. However if most of the time the label is simply the property key as title or sentence case, specifying a label decorator on every property feels like unnecessary boilerplate and doesn't make for a great developer experience.

If we had the ability to register a label provider function once at app startup, developers would be spared from having to specify labels for every field.

Something like this:

```typescript
import { registerLabelProvider } from 'tsdv-joi';
import * as Case from 'case';

registerLabelProvider(propertyName => Case.sentence(propertyName));
```

### LINKED ISSUE #41: Creating isolated instances of tsdv-joi
Currently the instance of `joi` used by `tsdv-joi` can be overridden by the application developer. However this is currently a global, which means the application could be using a library that also uses `tsdv-joi` and expects different behaviour from the `joi` instance or sets the instance itself.

In order to work better with other libraries, as @laurence-myers suggested, it would be better if consumers could create their own instances of `tsdv-joi` where they could set any specific configurations they need in complete isolation. As well as setting the `joi` instance, this would also be useful for registering a [label provider function](https://github.com/laurence-myers/tsdv-joi/issues/40) in isolation.

Existing exported decorators would continue to work as is with a *default* instance of `tsdv-joi`.

For example:

```typescript
import { TsdvJoi } from 'tsdv-joi';
import * as joi from 'joi';
import * as Case from 'case';

const tsdvJoi = new TsdvJoi({
  joi,
  labelProvider: propertyName => Case.sentence(propertyName)
});
```

### PR DESCRIPTION
**Pull request by codeandcats**
### Summary
This is a major rewrite of `tsdv-joi` which includes major breaking changes. The motivation is to allow for:
- Library/App developers to be able to create their own isolated versions of Joiful with different options (joi instance, options, label provider, etc) that will not conflict with each other

- Simplify scoping of decorators. Now all decorators for strings are called by first decorating your property with the `string` decorator and chaining the decorators, e.g. `@jf.string().min(8).max(20)`. This avoids incorrectly importing the wrong decorator for your property type (number min vs string min, etc) in the first place (less compiler errors). This also neatly scopes decorators to enable the path for default instance vs multiple instances of Joiful.

### Changes
- All root level decorators are now all scoped under a new `Joiful` class.

- Consumers can now import decorators directly from joiful, or create their own Joiful instance with different options (e.g. a label provider, different joi instance with different options, etc)

- Removed most object decorators to keep things simple for now.

- Typecheck, linting and tests now performed via `yarn run check`, so tests can be run separate to typecheck and linting. CI now runs `check` script.

- Added JSDoc comments to all decorators that were missing them

- Moved `build.ts` out of `src` into it's own `support` directory

- Renamed `src/main.ts` to `src/index.ts`

- 💯% test coverage across entire library

- Removed unused code

- Restrict (at runtime) custom joi instances to be the same major version that joiful expects (v10 at the moment)

Closes #23 and closes #24 and closes #40 and closes #41.

### CODE REVIEW

File: src/core.ts
Comment (laurence-myers) [inline src/core.ts]: ⛏ `options` could default to ` = {}`, then you don't need to check if `options` is undefined.

```typescript
export const getJoi = (options: { joi?: typeof Joi } = {}) => options.joi || Joi;
```

**Comment by laurence-myers** [comment]
The build is failing on CircleCI due to TS compilation failures, some calls to `getJoiSchema()` are not providing the Joi instance argument.

https://circleci.com/gh/joiful-ts/joiful/65

~~(We should see why the CI check is not being reported in the PR)~~

**Comment by codeandcats** [comment]
Interestingly I don't get the error when running tests locally. Though I can see that the jest config is pointing to the wrong tsconfig file, when I update it to point to the tsconfig for the tests, the tests will error locally.

I will hold off pushing the fixes until we sort out the CI integration. It might help to blow away the old Circle integration and reconfigure it under the new org?

**Comment by codeandcats** [comment]
Oh cool, looks like you fixed the integration, thanks! 👌

I'll address those issues.

File: src/joiful.ts
Comment (laurence-myers) [inline src/joiful.ts]: ⛏ Could be written in another way:

```typescript
constructor(private readonly options: JoifulOptions = {}) {}
```

This avoids the separate property definition, manually setting `this.options`, and needing to write `options || foo`. It does mean we set the property before calling `checkJoiIsCompatible()`, but since that function throws an error immediately it shouldn't matter.

File: src/validation.ts
Comment (laurence-myers) [inline src/validation.ts:86]: ⛏ I don't think you need to specify `| undefined` if you provide a default arg value.

File: src/validation.ts
Comment (laurence-myers) [inline src/validation.ts]: 🙏 I think the types here should be similar to `validateAsClass()`; `target: Partial<TInstance>[]`, and return type is `ValidationResult<TInstance[]>`

File: src/decorators/string.ts
Comment (laurence-myers) [inline src/decorators/string.ts:148]: 🙏 When we eventually update the README.md, let's remember to document any decorators that diverge from diverge from the Joi API. For example, we need to let consumers know to use `exactLength()` instead of `length`.

**Review by laurence-myers — APPROVED** [review]
This is frickin' awesome! I love these changes! 🎉🎈👯🎈 🎉 

I just wanted to point out how amazing this, in particular, is:
joiful_100pc_code_coverage

I've only ever seen 100% code coverage as an aspiration, but never achieved. Well done! 🥇 

Now, due to the scope of the changes, I haven't performed a thorough analysis. But we can deal with any issues that come up later.

One thing I missed, is there an example showing how to write custom decorators using the new API?

Something to consider is how to communicate the API changes to users. Perhaps a migration guide? Also, we'll probably need to add a warning to the `tsdv-joi` package, letting users know that it's been renamed to `joiful`.

File: src/validation.ts
Comment (codeandcats) [reply src/validation.ts:86]: In this case I do because `this.defaultOptions` can be `undefined`.

**Comment by codeandcats** [comment]
Thanks @laurence-myers. I'm thrilled that you like the new changes. 😄 Thanks for taking time out of your weekend to look them over.

Yeah, I'm really happy we finally got to 100% test coverage too. That's the nice thing about libraries - it's actually an achievable outcome. You just reminded me to set the min coverage thresholds so that we can't commit/merge in changes without keeping 100% coverage. I've pushed this in now. Of course if you want to commit in code and write the tests after (what I like to call Test Delayed Development 😉) you can still commit with `--no-verify` and the CI checks should stop us from merging to master. 👌

Creating your own shorthand decorators is actually really easy now.

```typescript
import * as jf from 'joiful';

const password = () => jf.string()
  .min(8)
  .regex(/[a-z]/)
  .regex(/[A-Z]/)
  .regex(/[0-9]/)
  .required();
```
I've added an [example](https://github.com/joiful-ts/joiful/pull/54/commits/af3bb94c9dd6e387dfe16d6fba2db3aa4b7e9892#diff-a85fca6ccb55572b1cf69c2cca5655c9R112) in the examples tests, but this would be great to have in the README too.

Great idea about writing a migration guide. I agree, we should update `tsdv-joi`'s README and add a deprecation notice on npm with instructions to use joiful instead and link to the migration guide. I figure we can worry about this after joiful is published.

Alright, I might merge this in then and like you said, if we spot anything else we can deal with it as it comes up. 👌

Thanks again.

**Pull request merged**
